whisper.cpp

```diff
@@ -4082,6 +4096,109 @@ static void whisper_sequence_score(
     }
 }
 
+static bool whisper_kv_swap_fast(std::vector<int> & view, whisper_decoder src[], int & size) {
+    // (decoder->buffer->decoder or decoder->buffer + decoder->decoder)
+    std::unordered_set<int> two_copy; // decoder indices require two copies to safely modify KV caches
+    two_copy.reserve(size);
+
+    // (buffer->decoder or decoder->decoder)
+    std::unordered_set<int> one_copy; // decoder indices require one copy to safely modify KV caches
+    one_copy.reserve(size);
+
+    // (decoder<->decoder)
+    std::unordered_set<int> p_swap_set; // decoder indices able to swap KV-cache pointers
+    p_swap_set.reserve(size);
+    std::vector<whisper_pair<int, int>> p_swap_vec;
+    p_swap_vec.reserve(size);
+
+    for (int i = 0; i < size; i++) {
+        // zero-copy (no modification)
+        if (i == view[i] || view[i] < 0) {continue;}
+
+        bool is_one_copy = true;
+        // since we modify data sequentially, we only consider decoder indices after current index
+        for (int j = i + 1; j < size; j++) {
+            if (i == view[j]) {
+                two_copy.insert(i);
+                is_one_copy = false;
+                // detect symmetric diagram
+                if (j == view[i]) {
+                    p_swap_set.insert(i);
+                    p_swap_set.insert(j);
+                    p_swap_vec.emplace_back(i, j);
+                }
+                break;
+            }
+        }
+        if (is_one_copy) {
+            one_copy.insert(i);
+        }
+    }
+
+    // beam-search helpers
+    struct kv_buf {
+        std::vector<uint8_t> k;
+        std::vector<uint8_t> v;
+    };
+
+    static std::vector<kv_buf> kv_bufs(size);
+    static bool is_resized = false;
```

Comment:
This is not OK - different whisper context will now share the same KV buffers, so this breaks parallel usage.
These buffers have to become part of `whisper_state` instead.

Also, no need for `is_resized` flag. Just always resize. The resize will be noop if the new size is small or equal to the old size.

Reply:
> This is not OK - different whisper context will now share the same KV buffers, so this breaks parallel usage.

Could you clarify this a bit more, please? I don't fully understand what you mean.

> Also, no need for is_resized flag. Just always resize. The resize will be noop if the new size is small or equal to the old size.

Agree.